### ISSUE TITLE
MPNN does not work on classification tasks

### ISSUE THREAD

**Issue by vid33** [deepchem/deepchem#1234]
It seems that getting an MPNN model to converge on a classification tasks, e.g. on tox21, is extremely tricky. 

I've experimented with a number of things, like trying different activation functions, using the Weave gather layer instead of the more complicated set-to-set model used in MPNN. 

Convergence issues seem to be largely independent of the choice of learning rate, and even the number of recurrent layers.

It could be that the combination of recurrent layers with multi-task classification simply doesn't play well in this setting, or of course I might be missing something more obvious...

Any ideas and suggestions on what to experiment with would be very welcome, as I'm kind of stuck at this stage, would like to find a solution, and contribute an improvement in this direction if applicable.

Thanks,

Vid

**Comment by rbharath** [comment]
@miaecle has some experience getting MPNN to converge on MoleculeNet and might be able to help.

In general, I'd recommend checking out the MoleculeNet parameters for MPNN and seeing if those work for you.

**Comment by miaecle** [comment]
@vid33 Thanks for trying this! MPNN is first designed for QM9 tasks and hence does not support classification in our implementation. I am also pretty curious on how the model perform on our standard classification sets. Just to clarify, by convergence, do you mean that the model performance(loss) will fluctuate?

**Comment by vid33** [comment]
Thanks for your thoughts. Yes, the optimisation algorithm is pretty much doing nothing - the output is close to random, auc \approx 0.5 for all tasks, and the loss might decrease a little for an epoch or two, but then keeps oscillating. Either it's a silly mistake in my implementation, or there's a deeper reason related to the use of recurrent layers.

**Comment by miaecle** [comment]
@vid33 I see, that's weird. I will also try to take a look here. Similar thing has happened to weave before, which is solved by just reducing learning rate. Hyperparameter tuning here might be pretty tricky.

**Comment by vid33** [comment]
Yes, I agree about about the hyperparameter tuning - fwiw I've tried reducing the learning rate to a very small value 1e-5 and below to no avail.  As the output of the message passing+update layer has the same format as the output of weave layers, I also tried using weave_gather instead of the set-to-set layer. As this doesn't seem to make any difference, assuming no silly errors, the issue presumably has to do with the initial message passing + update phase.

Excuse the long cut and paste from my modified MPNN model class below, but it'll give an idea of the kinds of things I've been trying, and might help identify any obvious errors.

 def build_graph(self):
    # Build placeholders
    self.atom_features = Feature(shape=(None, self.n_atom_feat))
    self.pair_features = Feature(shape=(None, self.n_pair_feat))
    self.atom_split = Feature(shape=(None,), dtype=tf.int32)
    self.atom_to_pair = Feature(shape=(None, 2), dtype=tf.int32)

    message_passing = MessagePassing(
        self.T,
        message_fn='enn',
        update_fn='gru',
        n_hidden=self.n_hidden,
        in_layers=[self.atom_features, self.pair_features, self.atom_to_pair])

    atom_embeddings_dense = Dense(self.n_hidden, activation_fn=tf.nn.relu, in_layers=[message_passing])

    atom_embeddings = BatchNormalization(epsilon=1e-5, mode=1, in_layers=[atom_embeddings_dense])

    weave_gather = WeaveGather(
        self.batch_size,
        n_input=self.n_hidden,
        gaussian_expand=True,
        in_layers=[atom_embeddings, self.atom_split])

    #mol_embeddings = SetGather(
    #    self.M,
    #    self.batch_size,
    #    n_hidden=self.n_hidden,
    #    in_layers=[atom_embeddings, self.atom_split])

    #dense1 = Dense(
    #    out_channels=2 * self.n_hidden,
    #    activation_fn=tf.nn.relu,
    #    in_layers=[mol_embeddings])
    costs = []
    self.labels_fd = []
    for task in range(self.n_tasks):
      if self.mode == "classification":
        classification = Dense(
            out_channels=2, activation_fn=None, in_layers=[weave_gather])
        softmax = SoftMax(in_layers=[classification])
        self.add_output(softmax)

        label = Label(shape=(None, 2))
        self.labels_fd.append(label)
        cost = SoftMaxCrossEntropy(in_layers=[label, classification])
        costs.append(cost)
      if self.mode == "regression":
        regression = Dense(
            out_channels=1, activation_fn=None, in_layers=[dense1])
        self.add_output(regression)

        label = Label(shape=(None, 1))
        self.labels_fd.append(label)
        cost = L2Loss(in_layers=[label, regression])
        costs.append(cost)
    if self.mode == "classification":
      all_cost = Stack(in_layers=costs, axis=1)
    elif self.mode == "regression":
      all_cost = Stack(in_layers=costs, axis=1)
    self.weights = Weights(shape=(None, self.n_tasks))
    loss = WeightedError(in_layers=[all_cost, self.weights])
    self.set_loss(loss)

**Comment by miaecle** [comment]
@vid33 No problem, thanks for sharing! This is pretty helpful, I will try reproduce the model on tox21.
